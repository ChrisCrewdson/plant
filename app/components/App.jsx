const React = require('react');
const actions = require('../actions');
const PropTypes = require('prop-types');

class App extends React.Component {
  static contextTypes = {
    store: PropTypes.object.isRequired,
  };

  componentWillMount() {
    const { store } = this.context;
    const users = store.getState().get('users');
    if (!users || users.size === 0) {
      store.dispatch(actions.loadUsersRequest());
    }

    const locations = store.getState().get('locations');
    if (!locations || locations.size === 0) {
      store.dispatch(actions.loadLocationsRequest());
    }

    // const user = store.getState().get('user');
    // if(user && user.get('isLoggedIn', false)) {
    //   const locationIds = user.get('locationIds');
    //   if(!locationIds) {
    //     const userId = user.get('_id');
    //     store.dispatch(actions.loadLocationsRequest(userId));
    //   }
    // }
  }

  render() {
    return (
      <div className="react-root">
        {this.props.children}
      </div>
    );
  }
}

App.propTypes = {
  // eslint-disable-next-line react/forbid-prop-types
  children: PropTypes.object.isRequired,
};

module.exports = App;
