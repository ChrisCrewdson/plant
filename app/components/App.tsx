import React from 'react';
import PropTypes from 'prop-types';
import { actionFunc } from '../actions';

export default class App extends React.Component {
  // TODO: When tsc 3.7+ is in use remove the ! to see hint text on how to change this.
  context!: PlantContext;

  static propTypes = {
    // eslint-disable-next-line react/forbid-prop-types
    children: PropTypes.object.isRequired,
  };

  static contextTypes = {
    // eslint-disable-next-line react/forbid-prop-types
    store: PropTypes.object.isRequired,
  };

  // eslint-disable-next-line camelcase
  UNSAFE_componentWillMount() {
    const { store } = this.context;
    const { users = {}, locations = {} } = store.getState();

    // TODO: This will cause a problem for a non-initialized site
    // that has zero users and zero locations as these values
    // will be 0 which is falsy.
    const usersCount = Object.keys(users).length;
    const locationsCount = Object.keys(locations).length;

    if (!usersCount) {
      store.dispatch(actionFunc.loadUsersRequest());
    }

    if (!locationsCount) {
      store.dispatch(actionFunc.loadLocationsRequest());
    }
  }

  render() {
    const { children } = this.props;
    return (
      <div className="react-root">
        {children}
      </div>
    );
  }
}
