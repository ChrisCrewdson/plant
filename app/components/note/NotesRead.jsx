const CircularProgress = require('material-ui/CircularProgress').default;
const Immutable = require('immutable');
const metrics = require('../../libs/metrics');
const NoteRead = require('./NoteRead');
const Paper = require('material-ui/Paper').default;
const PropTypes = require('prop-types');
const React = require('react');

const { List } = Immutable;

class NotesRead extends React.Component {
  componentWillMount() {
    this.sortNotes();
  }

  componentWillReceiveProps(nextProps) {
    this.sortNotes(nextProps);
  }

  sortNotes(props = this.props) {
    const { notes, plant } = props;
    const noteIds = plant.get('notes', List());
    if (!(List.isList(noteIds) || Immutable.Set.isSet(noteIds))) {
      // console.error('Not a List or Set from plant.get notes:', props.plant, noteIds);
    }

    if (!noteIds.size) {
      return;
    }

    const sortedIds = noteIds.sort((a, b) => {
      const noteA = notes.get(a);
      const noteB = notes.get(b);
      if (noteA && noteB) {
        const dateA = noteA.get('date');
        const dateB = noteB.get('date');
        if (dateA === dateB) {
          return 0;
        }
        return dateA > dateB ? 1 : -1;
      }
      return 0;
    });

    this.setState({ sortedIds });
  }

  render() {
    const {
      dispatch,
      interim,
      isOwner,
      notes,
      plant,
      plants,
      user,
    } = this.props;
    const { sortedIds } = this.state || {};
    if (!sortedIds || !sortedIds.size) {
      return null;
    }

    const paperStyle = {
      backgroundColor: '#ddd',
      display: 'inline-block',
      margin: 20,
      padding: 20,
      width: '100%',
    };

    const metricNotes = metrics.notesToMetricNotes(sortedIds, notes);
    const renderedNotes = metricNotes.map((metricNote) => {
      const {
        noteId, note, sinceLast, change,
      } = metricNote;
      switch (metricNote.type) {
        case 'note':
          return (<NoteRead
            dispatch={dispatch}
            interim={interim}
            isOwner={isOwner}
            key={noteId}
            note={note}
            notes={notes}
            plant={plant}
            plants={plants}
            user={user}
          />);
        case 'since':
          return (
            <Paper key={`${noteId}-sincelast`} style={paperStyle} zDepth={1}>
              {sinceLast}
            </Paper>);
        case 'metric':
          return (
            <Paper key={`${noteId}-change`} style={paperStyle} zDepth={1}>
              {change}
            </Paper>);
        case 'unfound':
          return (<CircularProgress key={noteId} />);
        default:
          throw new Error(`Unknown note render type ${metricNote.type}`);
      }
    });

    return (
      <div>
        {renderedNotes}
      </div>
    );
  }
}

NotesRead.propTypes = {
  dispatch: PropTypes.func.isRequired,
  interim: PropTypes.shape({
    get: PropTypes.func.isRequired,
    getIn: PropTypes.func.isRequired,
  }).isRequired,
  isOwner: PropTypes.bool.isRequired,
  notes: PropTypes.shape({
    get: PropTypes.func.isRequired,
  }).isRequired,
  plant: PropTypes.shape({
    get: PropTypes.func.isRequired,
  }).isRequired,
  plants: PropTypes.shape({
    get: PropTypes.func.isRequired,
    filter: PropTypes.func.isRequired,
  }).isRequired,
  user: PropTypes.shape({ // Immutable.js Map
    get: PropTypes.func.isRequired,
  }).isRequired,
};

module.exports = NotesRead;
