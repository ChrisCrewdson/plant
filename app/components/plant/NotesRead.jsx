import * as actions from '../../actions';
import CircularProgress from 'material-ui/CircularProgress';
import React from 'react';
import NoteRead from './NoteRead';
import Paper from 'material-ui/Paper';

export default class NotesRead extends React.Component {

  render() {
    const noteIds = [...(this.props.plant && this.props.plant.notes || [])];
    if(!noteIds.length) {
      return null;
    }
    const {notes} = this.props;

    // Find unloaded notes
    const unloaded = noteIds.reduce((acc, noteId) => {
      if(!notes[noteId]) {
        acc.push(noteId);
      }
      return acc;
    }, []);
    if(unloaded.length) {
      this.props.dispatch(actions.loadNotesRequest(unloaded));
    }

    const sortedIds = noteIds.sort((a, b) => {
      const noteA = notes[a];
      const noteB = notes[b];
      if(noteA && noteB) {
        if(noteA.date.isSame(noteB.date)) {
          return 0;
        }
        return noteA.date.isAfter(noteB.date) ? 1 : -1;
      } else {
        return 0;
      }
    });

    const paperStyle = {
      backgroundColor: '#ddd',
      display: 'inline-block',
      margin: 20,
      padding: 20,
      width: '100%',
    };

    let lastNoteDate;
    const renderedNotes = sortedIds.reduce((acc, noteId) => {
      const note = notes[noteId];
      if(note) {
        const sinceLast = lastNoteDate ? `...and then after ${lastNoteDate.from(note.date, true)}` : '';
        if(sinceLast && !lastNoteDate.isSame(note.date)) {
          acc.push(
            <Paper key={noteId + '-sincelast'} style={paperStyle} zDepth={1}>
              {sinceLast}
            </Paper>
          );
        }
        lastNoteDate = note.date;
        acc.push(
          <NoteRead
            key={noteId}
            {...this.props}
            note={note}
          />
        );
      } else {
        acc.push(
          <CircularProgress key={noteId} />
        );
      }
      return acc;
    }, []);

    return (
      <div>
        {renderedNotes}
      </div>
    );

  }
}

NotesRead.propTypes = {
  dispatch: React.PropTypes.func.isRequired,
  interim: React.PropTypes.shape({
    note: React.PropTypes.shape({
      note: React.PropTypes.object.isRequired,
      plant: React.PropTypes.object.isRequired,
    })
  }).isRequired,
  isOwner: React.PropTypes.bool.isRequired,
  notes: React.PropTypes.object.isRequired,
  plant: React.PropTypes.object.isRequired,
};
