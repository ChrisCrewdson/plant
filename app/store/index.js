const { createStore, applyMiddleware } = require('redux');
const reducers = require('../reducers'); // combineReducers already called on reducers in her)e
const api = require('../middleware/api');
const { setupSubscribe: userSubscribe } = require('./user');
const Immutable = require('immutable');

let middleware = [api];

if (process.env.NODE_ENV !== 'production') {
  const logger = require('../middleware/logger');
  middleware = [logger, api];
}

// Add the api to the pipeline/chain
const createStoreWithMiddleware = applyMiddleware(...middleware)(createStore);

const store = createStoreWithMiddleware(reducers, Immutable.fromJS(window.__INITIAL_STATE__ || {}));

userSubscribe(store);

module.exports = store;
