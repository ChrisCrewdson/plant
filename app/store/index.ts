export {}; // To get around: Cannot redeclare block-scoped variable .ts(2451)

const { createStore, applyMiddleware } = require('redux');

const seamless = require('seamless-immutable').static;
const reducers = require('../reducers'); // combineReducers already called on reducers in her)e
const api = require('../middleware/api');
const { setupSubscribe: userSubscribe } = require('./user');

const middleware = [api];

if (process.env.NODE_ENV !== 'production') {
  // eslint-disable-next-line global-require
  const logger = require('../middleware/logger');
  middleware.unshift(logger);
}

// Add the api to the pipeline/chain
const createStoreWithMiddleware = applyMiddleware(...middleware)(createStore);

// @ts-ignore - __INITIAL_STATE__ is added on the server
const initialState = seamless.from(window.__INITIAL_STATE__ || {});
const store = createStoreWithMiddleware(reducers, initialState);

userSubscribe(store);

module.exports = store;
