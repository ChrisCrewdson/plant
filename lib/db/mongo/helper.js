const _ = require('lodash');

module.exports = class Helper {
  static removeEmpty(doc) {
    if (!doc) {
      return doc;
    }
    if (_.isArray(doc)) {
      return doc.map(item => Helper.removeEmpty(item));
    }
    return _.pickBy(doc, item => item !== '');
  }

  /**
   * Helper function to convert _id from MongoId to string. Used in reads
   * @param {object} obj - Object that might have an _id
   * @returns {object} - the same object with a converted _id
   */
  static convertIdToString(obj) {
    if (_.isArray(obj)) {
      return obj.map(Helper.convertIdToString);
    }
    if (obj && obj._id) {
      // eslint-disable-next-line no-param-reassign
      obj._id = obj._id.toString();
    }
    return obj;
  }
};
