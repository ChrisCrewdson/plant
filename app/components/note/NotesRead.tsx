import PropTypes from 'prop-types';
import React from 'react';
import { Dispatch } from 'redux';

import CircularProgress from '@material-ui/core/CircularProgress';
import Paper from '@material-ui/core/Paper';

import NoteEdit from './NoteEdit';
import NoteRead from './NoteRead';
import { notesToMetricNotes } from '../../libs/metrics';
import utils from '../../libs/utils';
import { PlantAction } from '../../../lib/types/redux-payloads';
import Markdown from '../common/Markdown';
import { UiNotes, UiInterimNote } from '../../../lib/db/mongo/model-note';
import { UiInterim } from '../../../lib/types/model-interim';

interface NotesReadProps {
  dispatch: Dispatch<PlantAction<any>>;
  interim: UiInterim;
  locationId: string;
  notes: UiNotes;
  plant: UiPlantsValue;
  plants: UiPlants;
  userCanEdit: boolean;
}

export default function NotesRead(props: NotesReadProps): JSX.Element | null {
  const {
    dispatch,
    interim,
    locationId,
    notes,
    plant,
    plants,
    userCanEdit,
  } = props;
  const { notes: noteIds = [] } = plant;

  const sortedIds = noteIds.length
    ? utils.sortNotes(noteIds, notes)
    : [];


  const interimNote = ((interim && interim.note
      && interim.note.note) || {}) as UiInterimNote;
  const { isNew, _id: interimNoteId } = interimNote;

  if (interimNoteId && userCanEdit && !isNew) {
    return (
      <NoteEdit
        dispatch={dispatch}
        interimNote={interimNote}
        plants={plants}
        locationId={locationId}
      />
    );
  }

  if (!sortedIds || !sortedIds.length) {
    return null;
  }

  const paperStyle: React.CSSProperties = {
    backgroundColor: '#ddd',
    display: 'inline-block',
    margin: 20,
    padding: 20,
    width: '100%',
  };

  const metricNotes = notesToMetricNotes(sortedIds, notes);
  const renderedNotes = metricNotes.map((metricNote) => {
    const {
      noteId, note, sinceLast, change, type: noteType,
    } = metricNote;

    if (noteType === 'note') {
      if (note) {
        return (
          <NoteRead
            dispatch={dispatch}
            userCanEdit={userCanEdit}
            key={noteId}
            note={note}
            plant={plant}
          />
        );
      }
      return null;
    }

    switch (metricNote.type) {
      case 'since':
        return (
          <Paper key={`${noteId}-sincelast`} style={paperStyle} elevation={5}>
            {sinceLast}
          </Paper>
        );
      case 'metric':
        return (
          <Paper key={`${noteId}-change`} style={paperStyle} elevation={5}>
            <Markdown markdown={change} />
          </Paper>
        );
      case 'unfound':
        return (<CircularProgress key={noteId} />);
      default:
        throw new Error(`Unknown note render type ${metricNote.type}`);
    }
  });

  return (
    <div>
      {renderedNotes}
    </div>
  );
}


NotesRead.propTypes = {
  dispatch: PropTypes.func.isRequired,
  interim: PropTypes.shape({
    note: PropTypes.object,
  }).isRequired,
  locationId: PropTypes.string.isRequired,
  notes: PropTypes.shape({
    // _id: PropTypes.string.required,
  }).isRequired,
  plant: PropTypes.shape({
    notes: PropTypes.array,
  }).isRequired,
  plants: PropTypes.shape({
    notes: PropTypes.array,
  }).isRequired,
  userCanEdit: PropTypes.bool.isRequired,
};
