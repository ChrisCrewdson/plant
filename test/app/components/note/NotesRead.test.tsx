
import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { Dispatch } from 'redux';

import { MuiThemeProvider } from '@material-ui/core/styles';

import NotesRead from '../../../../app/components/note/NotesRead';
import store from '../../../../app/store';
import App from '../../../../app/components/App';
import { PlantAction } from '../../../../lib/types/redux-payloads';
import { theme } from '../../../helper';
import { UiNotes } from '../../../../lib/db/mongo/model-note';

describe('NotesRead', () => {
  const storeDispatch = store.dispatch;

  beforeAll(() => {
    store.dispatch = (() => {}) as Dispatch<PlantAction<any>>;

    // Date.parse('01 Jun 2019 00:00:00 GMT');
    // 1559347200000
    // Make the current Date.now a fixed known date so the result is consistent
    // because the component uses Date.now() (via Moment) for the time elapsed.
    const fakeCurrentDateTime = 1559347200000;
    Date.now = jest.fn(() => fakeCurrentDateTime);
  });

  afterAll(() => {
    store.dispatch = storeDispatch;
  });

  test('should render a NotesRead', () => {
    const interim = {
    };
    const locationId = 'l-1';
    const notes = {
      'n-1': {
        _id: 'n-1',
        date: 20190101,
        note: 'This is note text',
        plantIds: ['p-1'],
        title: 'Note #1',
      },
    } as unknown as UiNotes;
    const plant = {
      _id: 'p-1',
      notes: ['n-1'],
    } as unknown as UiPlantsValue;

    const plants: UiPlants = {
      'p-1': plant,
    };

    const component = renderer.create(
      <MuiThemeProvider theme={theme}>
        <Provider store={store}>
          <App>
            <MemoryRouter>
              <NotesRead
                dispatch={storeDispatch}
                interim={interim}
                locationId={locationId}
                notes={notes}
                plant={plant}
                plants={plants}
                userCanEdit
              />
            </MemoryRouter>
          </App>
        </Provider>
      </MuiThemeProvider>);
    const badgeTotals = component.toJSON();
    expect(badgeTotals).toMatchSnapshot();
  });
});
