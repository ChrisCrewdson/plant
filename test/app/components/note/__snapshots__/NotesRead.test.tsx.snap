// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotesRead should render a NotesRead 1`] = `
<div
  className="react-root"
>
  <div>
    <div
      className="MuiPaper-root MuiPaper-elevation5 MuiPaper-rounded"
      style={
        Object {
          "display": "inline-block",
          "margin": 20,
          "padding": 20,
          "width": "100%",
        }
      }
    >
      <div
        id="n-1"
      >
        <a
          href="?noteid=n-1#n-1"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M3.9 12c0-1.71 1.39-3.1 3.1-3.1h4V7H7c-2.76 0-5 2.24-5 5s2.24 5 5 5h4v-1.9H7c-1.71 0-3.1-1.39-3.1-3.1zM8 13h8v-2H8v2zm9-6h-4v1.9h4c1.71 0 3.1 1.39 3.1 3.1s-1.39 3.1-3.1 3.1h-4V17h4c2.76 0 5-2.24 5-5s-2.24-5-5-5z"
            />
          </svg>
        </a>
      </div>
      <h5>
        01-Jan-2019 (5 months ago)
      </h5>
      <div
        dangerouslySetInnerHTML={
          Object {
            "__html": "<p>This is note text</p>
",
          }
        }
      />
      <h2
        className="vcenter"
      >
        <div
          style={
            Object {
              "textAlign": "right",
            }
          }
        >
          <button
            className="MuiButtonBase-root MuiFab-root MuiFab-sizeMedium MuiFab-primary"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            title="Edit"
            type="button"
          >
            <span
              className="MuiFab-label"
            >
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root"
                focusable="false"
                role="presentation"
                style={
                  Object {
                    "fontSize": "3em",
                  }
                }
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                />
              </svg>
            </span>
            <span
              className="MuiTouchRipple-root"
            />
          </button>
          <button
            className="MuiButtonBase-root MuiFab-root MuiFab-sizeMedium MuiFab-secondary"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            style={
              Object {
                "marginLeft": "10px",
              }
            }
            tabIndex={0}
            title="Delete"
            type="button"
          >
            <span
              className="MuiFab-label"
            >
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root"
                focusable="false"
                role="presentation"
                style={
                  Object {
                    "fontSize": "3em",
                  }
                }
                viewBox="0 0 24 24"
              >
                <path
                  d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                />
              </svg>
            </span>
            <span
              className="MuiTouchRipple-root"
            />
          </button>
        </div>
      </h2>
    </div>
  </div>
</div>
`;
